### PR TITLE
Agent context storage wrapper should not override other wrappers

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7344

### CODE REVIEW

**Comment by oleborne** [comment]
Thanks @laurit for the fix. I was about to make the same proposal.

However this leaves indeed the possibility for the wrapper to break the agent by actually replacing the underlying `ContextStorage`. Such a wrapper would break the javaagent. 

If the code owners are ok with this possible issue, I would recommend to also update the Javadoc of `ContextStorageWrappersInstrumentation` to explicitly state that Wrappers are still allowed: only providers are bypassed.

**Comment by laurit** [comment]
> However this leaves indeed the possibility for the wrapper to break the agent by actually replacing the underlying `ContextStorage`. Such a wrapper would break the javaagent.

In my opinion this is not a big issue. If user has built such a wrapper then user can also modify it if there is desire to run with javaagent. If such a wrapper is used by a framework then agent can try to instrument it.

**Comment by oleborne** [comment]
> If user has built such a wrapper then user can also modify it if there is desire to run with javaagent.

Totally agree, hence my remark about updating the Javadoc to 1. warn people about the possibility and 2. clarify how to fix it should you be the author of such a wrapper.

File: instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/ContextStorageWrappersInstrumentation.java [hunk 1]
```diff
@@ -46,7 +46,7 @@ public static void methodExit(
         @Advice.Return(readOnly = false)
             List<Function<? super ContextStorage, ? extends ContextStorage>> wrappers) {
       wrappers = new ArrayList<>(wrappers);
-      wrappers.add(AgentContextStorage.wrap());
+      wrappers.add(0, AgentContextStorage.wrap());
```
Comment (trask) [inline instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/ContextStorageWrappersInstrumentation.java:49]:
```suggestion
      // AgentContextStorage wrapper doesn't delegate, so needs to be the innermost wrapper
      wrappers.add(0, AgentContextStorage.wrap());
```

**Review by trask — APPROVED** [review]

**Review by mateuszrzeszutek — APPROVED** [review]

**Pull request merged**